haystack/schema.py

```diff
@@ -1440,32 +1440,69 @@ def _build_document_metrics_df(
-            relevance_criterion_ids = list(query_df[query_df[relevance_criterion_col] == 1]["document_id"].values)
-            num_relevants = len(set(gold_ids + relevance_criterion_ids))
-            num_retrieved_relevants = query_df[relevance_criterion_col].values.sum()
-            rank_retrieved_relevants = query_df[query_df[relevance_criterion_col] == 1]["rank"].values
-            avp_retrieved_relevants = [
-                query_df[relevance_criterion_col].values[: int(rank)].sum() / rank for rank in rank_retrieved_relevants
-            ]
-
-            avg_precision = np.sum(avp_retrieved_relevants) / num_relevants if num_relevants > 0 else 0.0
-            recall_multi_hit = num_retrieved_relevants / num_relevants if num_relevants > 0 else 1.0
-            recall_single_hit = min(num_retrieved_relevants, 1) if num_relevants > 0 else 1.0
-            precision = num_retrieved_relevants / retrieved if retrieved > 0 else 0.0
-            rr = 1.0 / rank_retrieved_relevants.min() if len(rank_retrieved_relevants) > 0 else 0.0
-            dcg = (
-                np.sum([1.0 / np.log2(rank + 1) for rank in rank_retrieved_relevants])
-                if len(rank_retrieved_relevants) > 0
-                else 0.0
-            )
-            idcg = (
-                np.sum([1.0 / np.log2(rank + 1) for rank in range(1, num_relevants + 1)]) if num_relevants > 0 else 1.0
-            )
-            ndcg = dcg / idcg
+            relevant_rows = query_df[query_df[relevance_criterion_col] == 1]
+
+            relevance_criterion_ids = list(relevant_rows["document_id"].values)
+            if "document_id" in document_relevance_criterion:
+                # If it's a document_relevance_criterion that includes document_ids, use both, gold_document_ids and relevance_criterion_ids.
+                # If 'document_id' is used as document_relevance_criterion, relevance_criterion_ids is a subset of gold_document_ids.
+                gold_document_ids = list(query_df["gold_document_ids"].iloc[0])
+                num_relevants = len(set(gold_document_ids + relevance_criterion_ids))
+            else:
+                # Omit gold_document_ids if it is not used in the document_relevance_criterion
+                num_relevants = len(set(relevance_criterion_ids))
+                # num_relevants can be 0 even though there are relevant documents in the corpus, just because no relevant document has been retrieved.
+                # We fix this here by adding 1 if we know that this is not a no_answer query.
+                is_no_answer_query = query_df["gold_document_ids"].iloc[0] == []
+                if num_relevants == 0 and not is_no_answer_query:
+                    num_relevants = 999
+
+            num_retrieved = len(
+                query_df["document_id"]
+            )  # cannot be 0 as we have at least one (empty) document in the query_df
+            num_retrieved_relevants = relevant_rows[relevance_criterion_col].values.sum()
```

Comment:
Can be simplified by:
```suggestion
            num_retrieved_relevants = len(relevant_rows)
```